test/models/user_test.exs

```diff
@@ -0,0 +1,86 @@
+defmodule Yggdrasil.UserTest do
+  use Yggdrasil.ModelCase
+
+  alias Yggdrasil.User
+
+  @valid_attrs %{username: "tester", password: "password", password_confirmation: "password"}
+  @valid_attrs_upcase_user %{username: "TESTER", password: "password", password_confirmation: "password"}
+
+  @missing_username %{password: "password", password_confirmation: "pass"}
+  @missing_password %{username: "tester", password_confirmation: "password"}
+  @missing_password_conf %{username: "tester", password: "password"}
+  @password_mismatch %{username: "tester", password: "test", password_confirmation: "tester"}
+  @password_bad_len %{username: "tester", password: "te", password_confirmation: "tester"}
+  @password_conf_bad_len %{username: "tester", password: "te", password_confirmation: "te"}
+
+  @blank_msg "can't be blank"
+  @doesntmatch_msg "does not match confirmation"
+  @unique_msg "has already been taken"
+
+  @invalid_length_msg "should be at least %{count} character(s)"
+  @min_count 4
+
+  test "changeset with valid attributes" do
+    changeset = User.changeset(%User{}, @valid_attrs)
+    assert changeset.valid?
+  end
+
+  test "changeset with missing username" do
+    changeset = User.changeset(%User{}, @missing_username)
+
+    assert changeset.errors[:username] == @blank_msg
+    refute changeset.valid?
+  end
+
+  test "changeset with missing password" do
+    changeset = User.changeset(%User{}, @missing_password)
+
+    assert changeset.errors[:password] == @blank_msg
+    refute changeset.valid?
```

Comment:
I think I have solution to this, it may actually need to be like

create_changeset
update_password_changeset
update_changeset

at the moment update_changeset needs to verfiy that the hash was not changed and that username is included in the changeset

update_password_changeset needs to ensure username didn't change (feel like you won't be doing both at the same time) can drop the unique name constraint as it didn't change thus won't fire anyway and then applies the normal password/password_confirmation stuff and generates a new password as last step. This ensures that a password is saved unlike the current approach.

create_changeset does what it does now plus generate the password at the end.

this will encapsulate all the changes to user to those three functions, and the update one can expand as we add fields and same for create